cli/command/service/opts.go

```diff
@@ -430,6 +432,30 @@ func newServiceOptions() *serviceOptions {
 func (opts *serviceOptions) ToService() (swarm.ServiceSpec, error) {
 	var service swarm.ServiceSpec
 
+	envVariables, err := runconfigopts.ReadKVStrings(opts.envFile.GetAll(), opts.env.GetAll())
+	if err != nil {
+		return service, err
+	}
+
+	entries := []string{}
+	// Remove any duplicates by key ("K=V1 K=V2" => "K=V2").
+	// Note later values will override former values. The order remain unchanged.
+	// We iterate from the last.
+	envMap := map[string]bool{}
+	for i := len(envVariables) - 1; i >= 0; i-- {
+		kv := strings.SplitN(envVariables[i], "=", 2)
+		if !envMap[kv[0]] {
+			entries = append(entries, envVariables[i])
+		}
+		envMap[kv[0]] = true
+	}
+
+	// Need to reverse to restore the order
+	env := []string{}
+	for i := len(entries) - 1; i >= 0; i-- {
+		env = append(env, entries[i])
+	}
+
```

Comment:
We discussed this and think it's better to sort by key, and not trying to preserve the original sort order. This would make sure that the env-vars are always sorted the same way (and duplicates are already removed earlier, so that's no problem)